### ISSUE TITLE
windows_share property validation incorrectly reports 'path' missing

### ISSUE THREAD

**Issue by trickyearlobe** [chef-cookbooks/windows#490]
### Cookbook version
Fails in 3.1.1
Works in 3.1.0

### Chef-client version
12.9.38 and others

### Platform Details
Windows 2012r2 and others

### Scenario:
Using `windows_share` resource to create a share.

### Steps to Reproduce:
```
directory 'c:\testdir' do
  action :create
end

windows_share 'testdir' do
  action :create
  path 'c:\testdir'
  full_users ['everyone']
end
```

### Expected Result:
Share should be created

### Actual Result:

Validation incorrectly reports 'path' property missing with error below.
See also #482 and comments on commit fc2691ff5b126a8412a6caea2a5c99b6051ea54f

```
================================================================================
Error executing action `create` on resource 'windows_share[testdir]'
================================================================================

Chef::Exceptions::ValidationFailed
----------------------------------
path is required

Resource Declaration:
---------------------
# In C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/windows_share_test/recipes/default.rb

 23: windows_share 'testdir' do
 24:   action :create
 25:   path 'c:\testdir'
 26:   full_users ['everyone']
 27: end

Compiled Resource:
------------------
# Declared in C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/windows_share_test/recipes/default.rb:23:in `from_file'

windows_share("testdir") do
  action [:create]
  retries 0
  retry_delay 2
  default_guard_interpreter :default
  declared_type :windows_share
  cookbook_name "windows_share_test"
  recipe_name "default"
  path "c:\\testdir"
  full_users ["everyone"]
  share_name "testdir"
end

Platform:
---------
x64-mingw32
```

**Comment by lamont-granquist** [comment]
Can you post contents of the stacktrace.out file so we can see the full stack of where the exception is getting thrown?

**Comment by trickyearlobe** [comment]
@lamont-granquist here you go...

```
Generated at 2017-08-14 09:36:52 -0700
Chef::Exceptions::ValidationFailed: windows_share[testdir] (windows_share_test::default line 23) had an error: Chef::Exceptions::ValidationFailed: path is required
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/property.rb:368:in `get'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/property.rb:262:in `call'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/property.rb:512:in `path'
C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/windows/resources/share.rb:143:in `different_path?'
C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/windows/resources/share.rb:39:in `block in class_from_file'
(eval):2:in `block in action_create'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/provider.rb:360:in `instance_eval'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/provider.rb:360:in `compile_and_converge_action'
(eval):2:in `action_create'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/provider.rb:144:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/resource.rb:596:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/runner.rb:69:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/runner.rb:97:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/runner.rb:97:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/runner.rb:97:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/resource_collection/resource_list.rb:84:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/resource_collection/resource_list.rb:82:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/runner.rb:96:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/client.rb:667:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/client.rb:662:in `catch'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/client.rb:662:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/client.rb:701:in `converge_and_save'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/client.rb:281:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/application.rb:267:in `run_with_graceful_exit_option'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/application.rb:243:in `block in run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/local_mode.rb:44:in `with_server_connectivity'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/application.rb:226:in `run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/application/client.rb:456:in `sleep_then_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/application/client.rb:443:in `block in interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/application/client.rb:442:in `loop'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/application/client.rb:442:in `interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/application/client.rb:426:in `run_application'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/lib/chef/application.rb:58:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38-universal-mingw32/bin/chef-client:26:in `<top (required)>'
C:/opscode/chef/bin/chef-client:61:in `load'
C:/opscode/chef/bin/chef-client:61:in `<main>'
```

**Comment by trickyearlobe** [comment]
Note: I added a PR #491  to fix this... it reverts a previous commit that changes the way validation is done.

**Comment by lamont-granquist** [comment]
Yeah but the code as written in the patch you reverted is entirely correct on the face of it, so either core chef is busted, and we need to figure out why, or else there's some other bad code in this resource which is causing this bug which your patch does not address.

**Comment by lamont-granquist** [comment]
its blowing up on current_resource.path not being set not new_resource.path not being set:

https://github.com/chef-cookbooks/windows/blob/master/resources/share.rb#L143

**Comment by lamont-granquist** [comment]
this is enough to replicate it:

```ruby
% cat cookbooks/test/resources/stuff.rb
provides :stuff

property :thing, String, name_property: true
property :path, String, required: true

load_current_value do |desired|
  path nil
end

action :doit do
  if current_resource.path.nil?
    puts "it should not have blown up"
  end
end
```

since path is now `required` its also required on current_resource, but there's a code block in load_current_value that avoids ever setting that.  setting it to nil also does not set it...  and i think i'm using chef-13 so its not even a nillable thing there.

**Comment by trickyearlobe** [comment]
Not sure why it should be required on current resource... it should only be checking that we supplied a value for the desired state right? It's entirely possible that there is no existing path in the current state.

**Comment by lamont-granquist** [comment]
The code that checks the required stats is just on the property itself.  If its accessed and its set to required then the property throws the exception.  The property has no knowledge of if its hanging off of the new_resource or the current_resource.

It doesn't look like there's any sensible thing to set the current_resource.path to in this case since the array of returned values is empty because it hasn't ever been created.

One way to avoid that in this code is to set the `current_resource.path desired.path` and treat it like an identity attribute if the thing doesn't exist.  That will break the later `current_resource.path.nil?` check to check for the path already existing.

Another possibility is to expose `existing_share?` (probably expose `existing_share` as a memoized helper method, along with `existing_share?` as `!!existing_share`) and then use that instead of `current_resource.path.nil?` and not use the current_resource.path as a way to determine if there is an existing share or not.

**Comment by lamont-granquist** [comment]
@tas50 i'm feeling ambiguous about how to fix this in this cookbook right now, and the revert is simpler.  Want to hear what @coderanger thinks about the "right" way this should get solved and if we should go down the road of disabling these checks for current_resource's in core since I'm leaning that direction as well.

**Comment by tas50** [comment]
I'm fine with the revert. It's less than ideal, but we can live.

**Comment by tas50** [comment]
This should be fixed in 3.1.2
